bootstrap/api/v1beta1/conversion.go

```diff
@@ -16,79 +16,135 @@ package v1beta1
 import (
 	"fmt"
 
-	"sigs.k8s.io/controller-runtime/pkg/conversion"
+	"k8s.io/apimachinery/pkg/conversion"
+	utilconversion "sigs.k8s.io/cluster-api/util/conversion"
+	ctrlconversion "sigs.k8s.io/controller-runtime/pkg/conversion"
 
-	cabp3v1 "github.com/k3s-io/cluster-api-k3s/bootstrap/api/v1beta2"
+	bootstrapv1beta2 "github.com/k3s-io/cluster-api-k3s/bootstrap/api/v1beta2"
 )
 
 // ConvertTo converts the v1beta1 KThreesConfig receiver to a v1beta2 KThreesConfig.
-func (c *KThreesConfig) ConvertTo(dstRaw conversion.Hub) error {
-	dst := dstRaw.(*cabp3v1.KThreesConfig)
-	if err := autoConvert_v1beta1_KThreesConfig_To_v1beta2_KThreesConfig(c, dst, nil); err != nil {
+func (c *KThreesConfig) ConvertTo(dstRaw ctrlconversion.Hub) error {
+	dst := dstRaw.(*bootstrapv1beta2.KThreesConfig)
+	if err := Convert_v1beta1_KThreesConfig_To_v1beta2_KThreesConfig(c, dst, nil); err != nil {
 		return fmt.Errorf("converting KThreesConfig v1beta1 to v1beta2: %w", err)
 	}
+
+	restored := &bootstrapv1beta2.KThreesConfig{}
+	if ok, err := utilconversion.UnmarshalData(c, restored); err != nil {
+		return fmt.Errorf("unmarshalling stored conversion data: %w", err)
+	} else if !ok {
+		// No stored data.
+		return nil
+	}
+
+	dst.Spec.ServerConfig.CloudProviderName = restored.Spec.ServerConfig.CloudProviderName
+	dst.Spec.ServerConfig.DeprecatedDisableExternalCloudProvider = restored.Spec.ServerConfig.DeprecatedDisableExternalCloudProvider
```

Comment:
`DeprecatedDisableExternalCloudProvider` is not used in k3s/config.go 
Maybe we can just remove this field, and convert v1beta1 `DisableExternalCloudProvider` value to `CloudProviderName` and `DisableCloudController`

Reply:
I believe this is still needed for conversion. The deprecation best practice is also to deprecate the field in the next API version, and finally deleted it in the next-next version.

Reply:
Still you are right this should be converted to the new fields, I missed that.

Reply:
Should be fixed now with explicit conversion:

```go
	if !in.DisableExternalCloudProvider {
		out.CloudProviderName = "external"
		out.DisableCloudController = true
	} else {
		out.CloudProviderName = ""
		out.DisableCloudController = false
	}
```